### PR TITLE
Allow alpn disabled

### PR DESCRIPTION
**Pull request by nfuden**
# Description
Allow for alpn on  downstreams to be set as empty (disallows alpn on envoy!)

# Context
Expose potential (if dangerous) functionality on envoy with regards to alpn

# Checklist:

- [x] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [x] If I updated APIs (our protos) or helm values, I ran `make -B install-go-tools generated-code` to ensure there will be no code diff
- [x] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/5518

### CODE REVIEW

File: projects/gloo/pkg/utils/ssl.go [hunk 1]
```diff
@@ -71,10 +72,16 @@ func (s *sslConfigTranslator) ResolveDownstreamSslConfig(secrets v1.SecretList,
 	if common.GetValidationContextType() != nil {
 		requireClientCert = &wrappers.BoolValue{Value: !dc.GetOneWayTls().GetValue()}
 	}
+
 	// default alpn for downstreams.
 	if len(common.GetAlpnProtocols()) == 0 {
 		common.AlpnProtocols = []string{"h2", "http/1.1"}
 	}
+	// allow override for advanced usage to set to a dangerous setting
+	if len(common.GetAlpnProtocols()) == 1 && common.GetAlpnProtocols()[0] == constants.AllowEnabled {
```
Comment (jenshu) [inline projects/gloo/pkg/utils/ssl.go:81]: nit: this can be an `else if`

File: projects/gloo/pkg/utils/ssl_test.go [hunk 2]
```diff
@@ -191,7 +191,12 @@ var _ = Describe("Ssl", func() {
 			Expect(err).NotTo(HaveOccurred())
 			Expect(cfg.CommonTlsContext.AlpnProtocols).To(Equal([]string{"h2", "http/1.1"}))
 		})
-
+		It("should set alpn empty for downstream config", func() {
+			downstreamCfg.AlpnProtocols = []string{"allow_empty"}
```
Comment (jenshu) [inline projects/gloo/pkg/utils/ssl_test.go:195]: nit: this can use the constant you defined

File: projects/gloo/api/v1/ssl.proto [hunk 3]
```diff
@@ -33,6 +33,7 @@ message SslConfig {
 
     // Set Application Level Protocol Negotiation
     // If empty, defaults to ["h2", "http/1.1"].
+    // As an advanced option you may use ["allow_empty"]
```
Comment (jenshu) [inline projects/gloo/api/v1/ssl.proto:36]: can you add a bit more detail here, i.e. use allow_empty in order to explicitly set it to []

Comment (nfuden) [reply projects/gloo/api/v1/ssl.proto]: good call, have updated

File: projects/gloo/pkg/utils/ssl_test.go (hunk 2 above)
Comment (nfuden) [reply projects/gloo/pkg/utils/ssl_test.go]: I admit to having a toss up there to see if it it specifies for usage but for maintainability it definitely makes sense. In other words have updated per comment.

File: projects/gloo/pkg/utils/ssl.go (hunk 1 above)
Comment (nfuden) [reply projects/gloo/pkg/utils/ssl.go]: Updated

File: projects/gloo/constants/values.go [hunk 4]
```diff
@@ -0,0 +1,5 @@
+package constants
+
+// AllowEmpty is a reusable term to avoid setting defaults.
+// This should be used where gloo decides to set common sense defaults but needs a way to specify a empty set.
+const AllowEnabled = "allow_empty"
```
Comment (jenshu) [inline projects/gloo/constants/values.go:5]: should this be `AllowEmpty`?

File: docs/content/reference/api/github.com/solo-io/gloo/projects/gloo/api/v1/ssl.proto.sk.md [hunk 5]
```diff
@@ -57,7 +57,7 @@ SslConfig contains the options necessary to configure a virtual host or listener
 | `sniDomains` | `[]string` | optional. the SNI domains that should be considered for TLS connections. |
 | `verifySubjectAltName` | `[]string` | Verify that the Subject Alternative Name in the peer certificate is one of the specified values. note that a root_ca must be provided if this option is used. |
 | `parameters` | [.gloo.solo.io.SslParameters](../ssl.proto.sk/#sslparameters) |  |
-| `alpnProtocols` | `[]string` | Set Application Level Protocol Negotiation If empty, defaults to ["h2", "http/1.1"]. |
+| `alpnProtocols` | `[]string` | Set Application Level Protocol Negotiation If empty, defaults to ["h2", "http/1.1"]. As an advanced option you may use ["allow_empty"]. |
```
Comment (jenshu) [inline docs/content/reference/api/github.com/solo-io/gloo/projects/gloo/api/v1/ssl.proto.sk.md:60]: think you need to regenerate this too

**Review by jenshu — APPROVED** [review]

**Pull request merged**